### PR TITLE
Add  buildProperties to project service configuration

### PR DESCRIPTION
**Pull request by alexfdezsauco**
Add buildProperties to project service configuration.

The service configuration could look like this:

    ...
    services:
    - name: projectA
      project: ProjectA/ProjectA.csproj
      buildProperties: 
      - name: Configuration
        value: Debug

### CODE REVIEW

**Comment by davidfowl** [comment]
Instead of build args it should be a dictionary of properties.

**Comment by alexfdezsauco** [comment]
Ok, I will implement this with a dictionary. But there are entries in the dictionary that need to be transtaled as build command arguments. In this case, I extracted the "dictionary" only for the supported build configuration from the build args.

A question: What to do with the all properties in the dictionary? I know what to do with the build Configuration. But the others?

**Comment by davidfowl** [comment]
> Ok, I will implement this with a dictionary. But there are entries in the dictionary that need to be transtaled as build command arguments. In this case, I extracted the "dictionary" only for the supported build configuration from the build args.

They can all be translated via /p:{Key}={Value}, it's just msbuild properties. You can choose to recognize some as first class but I wouldn't at the moment.

**Comment by alexfdezsauco** [comment]
I tried something different for the moment, added properties but only build Configuration as first class

**Comment by alexfdezsauco** [comment]
Also translated extra properties as /p:{Key}={Value}

File: src/Microsoft.Tye.Core/ConfigModel/ConfigService.cs [hunk 1]
```diff
@@ -24,5 +24,6 @@ public class ConfigService
         public List<ConfigVolume> Volumes { get; set; } = new List<ConfigVolume>();
         [YamlMember(Alias = "env")]
         public List<ConfigConfigurationSource> Configuration { get; set; } = new List<ConfigConfigurationSource>();
+        public List<ConfigConfigurationSource> Properties { get; set; } = new List<ConfigConfigurationSource>();
```
Comment (davidfowl) [inline src/Microsoft.Tye.Core/ConfigModel/ConfigService.cs:27]: Did you reuse configuration source because it exists?

**Comment by davidfowl** [comment]
This change needs a test, my guess is that this will fail with --docker as well.

File: src/Microsoft.Tye.Core/ConfigModel/ConfigService.cs (hunk 1 above)
Comment (alexfdezsauco) [reply src/Microsoft.Tye.Core/ConfigModel/ConfigService.cs:27]: Yes

Comment (davidfowl) [reply src/Microsoft.Tye.Core/ConfigModel/ConfigService.cs:27]: I wouldn't. I would be specific about the intent of this field. It should be called buildProperties and should be a BuildProperty object

Comment (alexfdezsauco) [reply src/Microsoft.Tye.Core/ConfigModel/ConfigService.cs:27]: Ok

Comment (alexfdezsauco) [reply src/Microsoft.Tye.Core/ConfigModel/ConfigService.cs]: Fixed

**Comment by alexfdezsauco** [comment]
> This change needs a test, my guess is that this will fail with --docker as well.

Will check this one.

**Comment by alexfdezsauco** [comment]
> This change needs a test, my guess is that this will fail with --docker as well.

Added support for --docker option, working on the test

File: src/Microsoft.Tye.Core/ApplicationFactory.cs [hunk 2]
```diff
@@ -90,16 +91,22 @@ public static async Task<ApplicationBuilder> CreateAsync(OutputContext output, F
...
                     }
 
                     var executable = new ExecutableServiceBuilder(configService.Name!, expandedExecutable)
                     {
                         Args = configService.Args,
-                        WorkingDirectory = configService.WorkingDirectory != null ?
-                        Path.GetFullPath(Path.Combine(builder.Source.Directory.FullName, Environment.ExpandEnvironmentVariables(configService.WorkingDirectory))) :
-                        workingDirectory,
+                        WorkingDirectory =
+                                             configService.WorkingDirectory != null
```
Comment (davidfowl) [inline src/Microsoft.Tye.Core/ApplicationFactory.cs:103]: What's up with this formatting here.

File: src/Microsoft.Tye.Core/ApplicationFactory.cs [hunk 3]
```diff
@@ -117,21 +124,12 @@ public static async Task<ApplicationBuilder> CreateAsync(OutputContext output, F
                 builder.Services.Add(service);
 
                 // If there are no bindings and we're in ASP.NET Core project then add an HTTP and HTTPS binding
-                if (configService.Bindings.Count == 0 &&
-                    service is ProjectServiceBuilder project2 &&
-                    project2.IsAspNet)
+                if (configService.Bindings.Count == 0 && service is ProjectServiceBuilder project2 && project2.IsAspNet)
```
Comment (davidfowl) [inline src/Microsoft.Tye.Core/ApplicationFactory.cs:127]: Why did the formatting change here?

File: src/Microsoft.Tye.Core/ProjectReader.cs [hunk 4]
```diff
@@ -11,6 +11,7 @@
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 using System.Runtime.Loader;
+using System.Text.RegularExpressions;
```
Comment (davidfowl) [inline src/Microsoft.Tye.Core/ProjectReader.cs:14]: nit: Remove

File: src/Microsoft.Tye.Hosting/ProcessRunner.cs [hunk 5]
```diff
@@ -51,12 +53,32 @@ private async Task BuildAndRunProjects(Application application)
...
                 string args;
+                var buildArgs = string.Empty;
                 string workingDirectory;
                 if (serviceDescription.RunInfo is ProjectRunInfo project)
                 {
                     path = project.RunCommand;
                     workingDirectory = project.ProjectFile.Directory.FullName;
                     args = project.Args == null ? project.RunArguments : project.RunArguments + " " + project.Args;
+
+                    foreach (var supportedProperty in BuildPropertiesToOptionsMap)
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/ProcessRunner.cs:64]: Why do you need this?

File: test/E2ETest/Infrastructure/ConditionalTheoryAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Xunit;
+using Xunit.Sdk;
+
+namespace E2ETest
+{
+    using System;
```
Comment (davidfowl) [inline test/E2ETest/Infrastructure/ConditionalTheoryAttribute.cs:11]: Fix the usings to be out of the namespace..

File: src/Microsoft.Tye.Hosting/ProcessRunner.cs (hunk 5 above)
Comment (alexfdezsauco) [reply src/Microsoft.Tye.Hosting/ProcessRunner.cs]: Just removed, I thought that some properties must be turned into command option. But is not required.

File: test/E2ETest/Infrastructure/ConditionalTheoryAttribute.cs (hunk 6 above)
Comment (alexfdezsauco) [reply test/E2ETest/Infrastructure/ConditionalTheoryAttribute.cs]: Fixed

File: src/Microsoft.Tye.Core/ApplicationFactory.cs (hunk 3 above)
Comment (alexfdezsauco) [reply src/Microsoft.Tye.Core/ApplicationFactory.cs]: Not sure, I don't remember touching such part. But also fixed back.

File: src/Microsoft.Tye.Core/ApplicationFactory.cs (hunk 2 above)
Comment (alexfdezsauco) [reply src/Microsoft.Tye.Core/ApplicationFactory.cs]: Again, not sure. But also fixed back.

File: src/Microsoft.Tye.Core/ApplicationFactory.cs [hunk 7]
```diff
@@ -59,16 +59,17 @@ public static async Task<ApplicationBuilder> CreateAsync(OutputContext output, F
...
 
                     await ProjectReader.ReadProjectDetailsAsync(output, project);
 
                     // We don't apply more container defaults here because we might need
                     // to prompt for the registry name.
-                    project.ContainerInfo = new ContainerInfo()
-                    {
-                        UseMultiphaseDockerfile = false,
-                    };
+                    project.ContainerInfo = new ContainerInfo() { UseMultiphaseDockerfile = false, };
```
Comment (davidfowl) [inline src/Microsoft.Tye.Core/ApplicationFactory.cs:72]: The reformatting changes make it really hard to tell what actually changed. In the future when you send more PRs (and please send more PRs these are great) I'd refrain from doing these changes as they are a bit distracting in the diff.

Comment (davidfowl) [reply src/Microsoft.Tye.Core/ApplicationFactory.cs:72]: No problem. Your contributing some great stuff! Keep it coming!

**Pull request merged**